src/audio/selector/selector_generic.c

```diff
@@ -53,21 +73,23 @@ static void sel_s16le_1ch(struct comp_dev *dev, struct audio_stream *sink,
 static void sel_s16le_nch(struct comp_dev *dev, struct audio_stream *sink,
 			  const struct audio_stream *source, uint32_t frames)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	int16_t *src;
-	int16_t *dest;
-	uint32_t i;
-	uint32_t j = 0;
-	uint32_t channel;
-
-	for (i = 0; i < frames; i++) {
-		for (channel = 0; channel < cd->config.in_channels_count;
-		     channel++) {
-			src = audio_stream_read_frag_s16(source, j);
-			dest = audio_stream_write_frag_s16(sink, j);
-			*dest = *src;
-			j++;
-		}
+	int16_t *src = source->r_ptr;
+	int16_t *dst = sink->w_ptr;
+	int bmax;
+	int b;
+	int bytes_copied = 0;
+	const int bytes_total = frames * audio_stream_frame_bytes(source);
+
+	while (bytes_copied < bytes_total) {
+		b = bytes_total - bytes_copied;
+		bmax = audio_stream_bytes_without_wrap(source, src);
+		b = MIN(b, bmax);
+		bmax = audio_stream_bytes_without_wrap(sink, dst);
+		b = MIN(b, bmax);
+		memcpy(dst, src, b);
+		src = audio_stream_wrap(source, src + (b >> BYTES_TO_S16_SAMPLES));
+		dst = audio_stream_wrap(sink, dst + (b >> BYTES_TO_S16_SAMPLES));
```

Comment:
not a huge optimisation, but since you're using `memcpy()` now, `src` and `dst` don't need to be 16-bit pointers. You could make them 8-bit pointers and avoid shifting here. It might seem a bit counterintuitive to some, that we use 8-bit pointers in an S16LE copying function, so, it's up to you whether to make this change. There are more cases like this in this PR, possibly also in other your copy optimisation PRs.

Reply:
Yep, that would allow in compiled code to go from bytes to samples to bytes. I didn't do it because I was wondering if int8_t would get criticism. Let's re-try with bytes, it's at least most optimal code.